pkg/dns/dns.go

```diff
@@ -0,0 +1,404 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+
+	"istio.io/pkg/monitoring"
+
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
+)
+
+// Based on istio-ecosystem/istio-coredns-plugin
+// Changes from original:
+// - runs inside istiod, using Istio main gRPC server
+// - instead of directly reading from K8S, use istiod store.
+// - removed "log" - switching to istio log.
+// - refactored Query, so both DNS native interface and coredns grpc plugin are implemented
+// - added parts of istio-ecosystem/dns-discovery, to provide in process DNS
+
+// TODO:
+// - add metrics, ideally using same names as kubedns/coredns ( Doug ?)
+// - config options on what suffix to capture in agent
+
+// IstioDNS exposes a DNS interface to internal Istiod service database.
+// This can be used:
+// - as a CoreDNS gRPC plugin
+// - as a DNS-over-TLS resolver, with support for forwarding to k8s or upstream
+// - for debug - a plain DNS-over-UDP interface.
+//
+// The code is currently targeted for Istiod, with a per/pod or per/VM coreDNS
+// forwarding to it, and using the same XDS grpc server and cert.
+//
+// In future we may embed it in istio-agent as well, using XDS to sync the config store.
+type IstioDNS struct {
+	mux *dns.ServeMux
+
+	// local DNS-UDP server.
+	// Active in agent and istiod.
+	server *dns.Server
+
+	// TODO: add a dns-over-TCP server and capture, for istio-agent.
+
+	// local DNS-TLS server. This is active only in istiod.
+	tlsServer *dns.Server
+
+	resolvConfServers []string
+
+	running bool
+
+	client *dns.Client
+
+	// tlsClient is a client initialized with DNS-over-TLS certificates.
+	// If nil, no DNS-TLS requests will be made.
+	tlsClient   *dns.Client
+	tlsUpstream string
+
+	// m protects pending, conn and outID
+	m       sync.Mutex
+	pending map[uint16]chan *dns.Msg
+	conn    *dns.Conn
+	// outID is used to match requests to responses in the DNS-TCP.
+	outID        uint16
+	dnsTLSSuffix []string
+}
+
+var (
+	// DNSAddr is the env controlling the DNS-over-TLS server.
+	// By default will be active, set to empty string to disable DNS functionality.
+	DNSAddr = env.RegisterStringVar("dnsAddr", ":15053", "DNS listen address")
+
+	// DNSAgentAddr is the listener address on the agent.
+	// This is in the range of hardcoded address used by agent - not customizable
+	// except for tests.
+	// By default will be active, set to empty string to disable DNS functionality.
+	// Iptables interception matches this.
+	DNSAgentAddr = ":15013"
+
+	// DNSUp controls the upstream DNS.
+	DNSUp = env.RegisterStringVar("dnsUp", "istiod.istio-system.svc:853",
```

Comment:
nit: name this DNS_UP

Reply:
can we default to istiod_addr? so we don't need to specify istiod, ca, and DNS, when they are all the same 99% of the time?

Reply:
And if not we need to change the templates to specify this or it will break for multiple revisions

Reply:
Good point. Will do it.